expression/builtin_compare.go

```diff
@@ -1023,81 +1024,131 @@ func isTemporalColumn(expr Expression) bool {
 }
 
 // tryToConvertConstantInt tries to convert a constant with other type to a int constant.
-func tryToConvertConstantInt(ctx sessionctx.Context, con *Constant) *Constant {
+func tryToConvertConstantInt(ctx sessionctx.Context, isUnsigned bool, con *Constant) (_ *Constant, isOverflow bool) {
 	if con.GetType().EvalType() == types.ETInt {
-		return con
+		return con, false
 	}
 	dt, err := con.Eval(nil)
 	if err != nil {
-		return con
+		return con, false
 	}
 	sc := ctx.GetSessionVars().StmtCtx
-	i64, err := dt.ToInt64(sc)
+	fieldType := types.NewFieldType(mysql.TypeLonglong)
+	if isUnsigned {
+		fieldType.Flag |= mysql.UnsignedFlag
+	}
+	dt, err = dt.ConvertTo(sc, fieldType)
 	if err != nil {
-		return con
+		return con, terror.ErrorEqual(err, types.ErrOverflow)
 	}
 	return &Constant{
-		Value:        types.NewIntDatum(i64),
-		RetType:      types.NewFieldType(mysql.TypeLonglong),
+		Value:        dt,
+		RetType:      fieldType,
 		DeferredExpr: con.DeferredExpr,
-	}
+	}, false
 }
 
 // RefineConstantArg changes the constant argument to it's ceiling or flooring result by the given op.
-func RefineConstantArg(ctx sessionctx.Context, con *Constant, op opcode.Op) *Constant {
+func RefineConstantArg(ctx sessionctx.Context, isUnsigned bool, con *Constant, op opcode.Op) (_ *Constant, isOverflow bool) {
 	dt, err := con.Eval(nil)
 	if err != nil {
-		return con
+		return con, false
 	}
 	sc := ctx.GetSessionVars().StmtCtx
-	i64, err := dt.ToInt64(sc)
+	intFieldType := types.NewFieldType(mysql.TypeLonglong)
+	if isUnsigned {
+		intFieldType.Flag |= mysql.UnsignedFlag
+	}
+	var intDatum types.Datum
+	intDatum, err = dt.ConvertTo(sc, intFieldType)
 	if err != nil {
-		return con
+		return con, terror.ErrorEqual(err, types.ErrOverflow)
 	}
-	datumInt := types.NewIntDatum(i64)
-	c, err := datumInt.CompareDatum(sc, &con.Value)
+	c, err := intDatum.CompareDatum(sc, &con.Value)
 	if err != nil {
-		return con
+		return con, false
 	}
 	if c == 0 {
 		return &Constant{
-			Value:        datumInt,
-			RetType:      types.NewFieldType(mysql.TypeLonglong),
+			Value:        intDatum,
+			RetType:      intFieldType,
 			DeferredExpr: con.DeferredExpr,
-		}
+		}, false
 	}
 	switch op {
 	case opcode.LT, opcode.GE:
 		resultExpr := NewFunctionInternal(ctx, ast.Ceil, types.NewFieldType(mysql.TypeUnspecified), con)
 		if resultCon, ok := resultExpr.(*Constant); ok {
-			return tryToConvertConstantInt(ctx, resultCon)
+			return tryToConvertConstantInt(ctx, isUnsigned, resultCon)
 		}
 	case opcode.LE, opcode.GT:
 		resultExpr := NewFunctionInternal(ctx, ast.Floor, types.NewFieldType(mysql.TypeUnspecified), con)
 		if resultCon, ok := resultExpr.(*Constant); ok {
-			return tryToConvertConstantInt(ctx, resultCon)
+			return tryToConvertConstantInt(ctx, isUnsigned, resultCon)
+		}
+	case opcode.NullEQ, opcode.EQ:
+		switch con.RetType.EvalType() {
+		// e.g. 1 = 1.1 or 1 <=> 1.1 should always be false.
+		case types.ETReal, types.ETDecimal:
```

Comment:
Will this make `1 = 1.0` failed?

Reply:
Noop. line 1067 will check cases like 1=1.0.